paddle/operators/detail/grpc_client.h

```diff
@@ -0,0 +1,269 @@
+    send_status_.end = send_status_.start;
+    send_status_.var.ep = ep_;
+    send_status_.var.name = name;
+
+    const time_point deadline =
+        send_status_.start + std::chrono::milliseconds(time_out);
+    context_->set_deadline(deadline);
+
+    sendrecv::VariableMessage req;
+    _Request(scope, name, &req);
+
+    rpc_ = stub_->AsyncSendVariable(context_.get(), req, cq_.get());
+    rpc_->Finish(reply_.get(), status_.get(), (void*)this);
+
+    return true;
+  }
+
+  bool ProcRet() { return true; }
+
+  grpc::Status* GetStatus() { return status_.get(); }
+
+  SendStatus GetSendStatus() { return send_status_; }
+
+ private:
+  void _Request(const framework::Scope* scope, const std::string name,
+                sendrecv::VariableMessage* msg) {
+    // FIXME(gongwb): pass device context to here.
+    auto ctx = platform::CPUDeviceContext();
+    auto* var = scope->FindVar(name);
+    PADDLE_ENFORCE(var);
+    // TODO(gongwb): support SelectedRows
+    PADDLE_ENFORCE(var->IsType<framework::LoDTensor>(),
+                   "Only support LoDTensor, %s has wrong type", name);
+
+    const framework::LoDTensor& tensor = var->Get<framework::LoDTensor>();
+    std::ostringstream oss;
+    framework::SerializeToStream(oss, tensor, ctx);
+    msg->set_varname(name);
+    msg->set_serialized(oss.str());
+  }
+
+ private:
+  std::shared_ptr<grpc::CompletionQueue> cq_;
+  std::shared_ptr<grpc::Channel> ch_;
+  std::string ep_;
+  std::unique_ptr<sendrecv::SendRecvService::Stub> stub_;
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::Status> status_;
+  std::unique_ptr<sendrecv::VoidMessage> reply_;
+  std::unique_ptr<grpc::ClientAsyncResponseReader<sendrecv::VoidMessage>> rpc_;
+  SendStatus send_status_;
+};
+
+class GetClient : public ClientBase {
+ public:
+  GetClient(std::shared_ptr<grpc::CompletionQueue> cq,
+            std::shared_ptr<grpc::Channel> ch, std::string ep) {
+    cq_ = cq;
+    ch_ = ch;
+    ep_ = ep;
+    stub_ = sendrecv::SendRecvService::NewStub(ch);
+    context_.reset(new grpc::ClientContext());
+    status_.reset(new grpc::Status());
+    reply_.reset(new sendrecv::VoidMessage());
+  }
+
+  virtual ~GetClient() {}
+
+  bool Call(const framework::Scope* scope, const std::string& name,
+            int64_t time_out) {
+    return true;
+  }
+
+  bool ProcRet() { return true; }
+
+  grpc::Status* GetStatus() { return status_.get(); }
+
+  SendStatus GetSendStatus() { return send_status_; }
+
+ private:
+  std::shared_ptr<grpc::CompletionQueue> cq_;
+  std::shared_ptr<grpc::Channel> ch_;
+  std::string ep_;
+  std::unique_ptr<sendrecv::SendRecvService::Stub> stub_;
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::Status> status_;
+  std::unique_ptr<sendrecv::VoidMessage> reply_;
+  std::unique_ptr<grpc::ClientAsyncResponseReader<sendrecv::VoidMessage>> rpc_;
+  SendStatus send_status_;
+};
+
+class RPCClients {
```

Comment:
I still believe that send_op should be responsible to determine which endpoint should each variable send to. Client is only responsible for how to do communication, send_op is responsible for how to manage the send, if we are going to implement something like "ring-based all reduce", will only need to write a new `send_op` and do not need to change this client implementation.

Reply:
Done.Thanks.